packages/flutter_tools/test/general.shard/asset_bundle_variant_test.dart

```diff
@@ -9,84 +9,146 @@ import 'package:file/memory.dart';
+import 'package:flutter_tools/src/base/platform.dart';
+import 'package:flutter_tools/src/base/user_messages.dart';
+import 'package:flutter_tools/src/cache.dart';
 
-import 'package:flutter_tools/src/globals.dart' as globals;
+import 'package:flutter_tools/src/project.dart';
 
 import '../src/common.dart';
-import '../src/context.dart';
 
 void main() {
-  String fixPath(String path) {
-    // The in-memory file system is strict about slashes on Windows being the
-    // correct way so until https://github.com/google/file.dart/issues/112 is
-    // fixed we fix them here.
-    // TODO(dantup): Remove this function once the above issue is fixed and
-    // rolls into Flutter.
-    return path.replaceAll('/', globals.fs.path.separator);
+
+  Future<Map<String, List<String>>> extractAssetManifestFromBundle(ManifestAssetBundle bundle) async {
+    final String manifestJson = utf8.decode(await bundle.entries['AssetManifest.json']!.contentsAsBytes());
+    final Map<String, dynamic> parsedJson = json.decode(manifestJson) as Map<String, dynamic>;
+    final Iterable<String> keys = parsedJson.keys;
+    final Map<String, List<String>> parsedManifest = <String, List<String>> {
+      for (final String key in keys) key: List<String>.from(parsedJson[key] as List<dynamic>),
+    };
+    return parsedManifest;
   }
 
   group('AssetBundle asset variants', () {
-    late FileSystem testFileSystem;
-    setUp(() async {
-      testFileSystem = MemoryFileSystem(
-        style: globals.platform.isWindows
-          ? FileSystemStyle.windows
-          : FileSystemStyle.posix,
-      );
-      testFileSystem.currentDirectory = testFileSystem.systemTempDirectory.createTempSync('flutter_asset_bundle_variant_test.');
-    });
+    testWithoutContext('Only images in folders named with device pixel ratios (e.g. 2x, 3.0x) should be considered as variants of other images', () async {
+      final MemoryFileSystem fs = MemoryFileSystem.test();
+
+      const String image = 'assets/image.jpg';
+      const String image2xVariant = 'assets/2x/image.jpg';
+      const String imageNonVariant = 'assets/notAVariant/image.jpg';
+      final List<String> assets = <String>[
+        image,
+        image2xVariant,
+        imageNonVariant
+      ];
 
-    testUsingContext('main asset and variants', () async {
-      globals.fs.file('pubspec.yaml')
-        ..createSync()
-        ..writeAsStringSync(
+      for (final String asset in assets) {
+        final File assetFile = fs.file(asset);
+        await assetFile.create(recursive: true);
+        await assetFile.writeAsString(asset);
+      }
+
+      fs.file('.packages').createSync();
+
+      fs.file('pubspec.yaml').writeAsStringSync(
 '''
 name: test
 dependencies:
   flutter:
     sdk: flutter
 flutter:
   assets:
-    - a/b/c/foo
+    - assets/
 '''
       );
-      globals.fs.file('.packages').createSync();
+
+      final Platform platform = FakePlatform();
+
+      Cache.flutterRoot = Cache.defaultFlutterRoot(
```

Comment:
Could you lift setting this to a `setUp()` call so that it is more obvious we're doing this in case it leads to test errors later on? Also, since it's just a String, I think you can just hard-code whatever `Cache.defaultFlutterRoot` is returning, rather than having to call it and lift the dependencies.

Reply:
Sure. I also lifted up some of the boilerplate file configuration.